theta_dup/include/theta_sketch_dup.h

```diff
@@ -0,0 +1,2209 @@
+
+  // TODO: support intersection
+  // friend theta_intersection_alloc<A>;
+  // TODO: support a_not_b
+  // friend theta_a_not_b_alloc<A>;
+  static inline uint32_t get_capacity(uint8_t lg_cur_size, uint8_t lg_nom_size);
+  static inline uint32_t get_stride(uint64_t hash, uint8_t lg_size);
+
+  /**
+   * search hash values, if exists increase the count and return true, otherwise
+   * insert the value and return false
+   * @param hash: the hash value
+   * @param table: the pointer to the hash table
+   * @param table: lg_size of the current hash table
+   */
+  static bool hash_search_or_insert(uint64_t hash,
+                                    std::pair<uint64_t, uint64_t>* table,
+                                    uint8_t lg_size);
+  /**
+   * search hash values, if exists decrease the count
+   * if the count==0, remove the element from hash table and return true else
+   * return false
+   * @param hash: the hash value
+   * @param table: the pointer to the hash table
+   * @param table: lg_size of the current hash table
+   */
+  static bool hash_search_or_remove(uint64_t hash,
+                                    std::pair<uint64_t, uint64_t>* table,
+                                    uint8_t lg_size);
+  static bool hash_search(uint64_t hash,
+                          const std::pair<uint64_t, uint64_t>* table,
+                          uint8_t lg_size);
+
+  friend theta_sketch_dup_alloc<A>;
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      std::istream& is, resize_factor rf, uint8_t lg_cur_size,
+      uint8_t lg_nom_size, uint8_t flags_byte, uint64_t seed);
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      datasketches_pb::Update_theta_sketch_dup& pb, resize_factor rf,
+      uint8_t lg_cur_size, uint8_t lg_nom_size, uint8_t flags_byte,
+      uint64_t seed);
+  static update_theta_sketch_dup_alloc<A> internal_deserialize(
+      const void* bytes, size_t size, resize_factor rf, uint8_t lg_cur_size,
+      uint8_t lg_nom_size, uint8_t flags_byte, uint64_t seed);
+};
+
+// compact sketch
+
+template <typename A>
+class compact_theta_sketch_dup_alloc : public theta_sketch_dup_alloc<A> {
+ public:
+  static const uint8_t SKETCH_TYPE = 3;
+
+  // No constructor here.
+  // Instances of this type can be obtained:
+  // - by compacting an update_theta_sketch_dup
+  // - as a result of a set operation
+  // - by deserializing a previously serialized compact sketch
+
+  compact_theta_sketch_dup_alloc(const theta_sketch_dup_alloc<A>& other,
+                                 bool ordered);
+  virtual ~compact_theta_sketch_dup_alloc() = default;
+
+  virtual uint32_t get_num_retained() const;
+  virtual uint16_t get_seed_hash() const;
+  virtual bool is_ordered() const;
+  virtual string<A> to_string(bool print_items = false) const;
+  virtual void serialize(std::ostream& os) const;
+  virtual void serialize(datasketches_pb::Update_theta_sketch_dup* pb) const;
```

Comment:
Declaration: serialize of another **derived class** compact_theta_sketch_dup_alloc using protobuf.